validator/src/state/merkle.rs

```diff
@@ -559,6 +561,54 @@ fn write_change_log(
     Ok(db_writer.index_put(CHANGE_LOG_INDEX, root_hash, &change_log.write_to_bytes()?)?)
 }
 
+fn increment_ref_count(
+    db_writer: &mut LmdbDatabaseWriter,
+    key: &[u8],
+) -> Result<u64, StateDatabaseError> {
+    let ref_count = get_ref_count(db_writer, key)?;
+
+    db_writer.index_put(DUPLICATE_LOG_INDEX, key, &to_bytes(ref_count + 1))?;
+
+    Ok(ref_count)
+}
+
+fn decrement_ref_count(
+    db_writer: &mut LmdbDatabaseWriter,
+    key: &[u8],
+) -> Result<u64, StateDatabaseError> {
+    let count = get_ref_count(db_writer, key)?;
+    Ok(if count == 1 {
+        db_writer.index_delete(DUPLICATE_LOG_INDEX, key)?;
+        0
+    } else {
+        db_writer.index_put(DUPLICATE_LOG_INDEX, key, &to_bytes(count - 1))?;
+        count - 1
+    })
+}
+
+fn get_ref_count(
+    db_writer: &mut LmdbDatabaseWriter,
+    key: &[u8],
+) -> Result<u64, StateDatabaseError> {
+    Ok(
+        if let Some(ref_count) = db_writer.index_get(DUPLICATE_LOG_INDEX, key)? {
+            from_bytes(ref_count)
+        } else {
+            0
+        },
+    )
+}
+
+fn to_bytes(num: u64) -> [u8; 8] {
+    unsafe { ::std::mem::transmute(num) }
```

Comment:
This should standardize on a memory representation before transmuting: https://doc.rust-lang.org/nightly/src/core/num/mod.rs.html#3706-3708 Same with from_bytes()